### PR TITLE
Configurable recent chat

### PR DESCRIPTION
**Pull request by EXio4**
Implements https://github.com/minetest/minetest/issues/6096

### CODE REVIEW

File: src/chat.cpp [hunk 1]
```diff
@@ -731,6 +738,11 @@ void ChatBackend::clearRecentChat()
 	m_recent_buffer.clear();
 }
 
+
+void ChatBackend::applySettings(Settings* settings) {
+	m_recent_buffer.resize(settings->getU32("recent_chat_size"));
```
Comment (red-001) [inline src/chat.cpp]: Why are you passing g_settings to this function instead of accessing it directly?

File: builtin/settingtypes.txt [hunk 2]
```diff
@@ -634,6 +634,9 @@ crosshair_color (Crosshair color) string (255,255,255)
 #    Crosshair alpha (opaqueness, between 0 and 255).
 crosshair_alpha (Crosshair alpha) int 255 0 255
 
+#    Maximum number of recent chat items to show
+recent_chat_size (Recent Chat Messages) int 6 3 99
```
Comment (red-001) [inline builtin/settingtypes.txt]: 99 seems a bit high, maybe set the max to something more reasonable like 20 ?

**Comment by paramat** [comment]
Better i think the setting refers 'lines' of chat instead of 'size'.

File: src/chat.cpp [hunk 3]
```diff
@@ -731,6 +738,11 @@ void ChatBackend::clearRecentChat()
 	m_recent_buffer.clear();
 }
 
+
+void ChatBackend::applySettings() {
```
Comment (nerzhul) [inline src/chat.cpp]: brace on wrong line

File: src/chat.cpp [hunk 4]
```diff
@@ -369,6 +369,13 @@ s32 ChatBuffer::getBottomScrollPos() const
 	return formatted_count - rows;
 }
 
+void ChatBuffer::resize(u32 scrollback) {
+	m_scrollback = scrollback;
+	if (m_unformatted.size() > m_scrollback)
```
Comment (nerzhul) [inline src/chat.cpp]: brace on wrong line

File: src/game.cpp [hunk 5]
```diff
@@ -2051,6 +2051,9 @@ bool Game::initGui()
 	// Remove stale "recent" chat messages from previous connections
 	chat_backend->clearRecentChat();
 
+    // Make sure the size of the recent messages buffer is right
+    chat_backend->applySettings();
```
Comment (nerzhul) [inline src/game.cpp]: wrong indent

File: src/settings_translation_file.cpp [hunk 6]
```diff
@@ -3,14 +3,13 @@
 // To update it, refer to the bottom of builtin/mainmenu/dlg_settings_advanced.lua
```
Comment (nerzhul) [inline src/settings_translation_file.cpp]: unrelated modification

File: src/chat.cpp (hunk 4 above)
Comment (sfan5) [reply src/chat.cpp]: braces can be removed actually

File: src/settings_translation_file.cpp (hunk 6 above)
Comment (sfan5) [reply src/settings_translation_file.cpp]: nah, it's okay to regenerate this
just don't squash it on merging

File: src/chat.cpp [hunk 7]
```diff
@@ -731,6 +739,12 @@ void ChatBackend::clearRecentChat()
 	m_recent_buffer.clear();
 }
 
+
+void ChatBackend::applySettings()
+{
+	m_recent_buffer.resize(g_settings->getU32("recent_chat_lines"));
```
Comment (SmallJoker) [inline src/chat.cpp]: Needs range limit

File: src/chat.cpp [hunk 8]
```diff
@@ -369,6 +369,14 @@ s32 ChatBuffer::getBottomScrollPos() const
 	return formatted_count - rows;
 }
 
+void ChatBuffer::resize(u32 scrollback)
+{
+	m_scrollback = scrollback;
+	if (m_unformatted.size() > m_scrollback)
+	{
```
Comment (SmallJoker) [inline src/chat.cpp]: Brace removal possible. Wrong style

File: src/chat.cpp [hunk 9]
```diff
@@ -369,6 +369,14 @@ s32 ChatBuffer::getBottomScrollPos() const
...
+		deleteOldest(m_unformatted.size() - m_scrollback);
+	}
```
Comment (sfan5) [inline src/chat.cpp]: l

File: builtin/settingtypes.txt [hunk 10]
```diff
@@ -634,6 +634,9 @@ crosshair_color (Crosshair color) string (255,255,255)
 #    Crosshair alpha (opaqueness, between 0 and 255).
 crosshair_alpha (Crosshair alpha) int 255 0 255
 
+#    Maximum number of recent chat items to show
```
Comment (paramat) [inline builtin/settingtypes.txt]: Should also say 'lines'.

File: src/chat.cpp [hunk 11]
```diff
@@ -731,6 +737,14 @@ void ChatBackend::clearRecentChat()
 	m_recent_buffer.clear();
 }
 
+
+void ChatBackend::applySettings()
+{
+	u32 recent_lines = g_settings->getU32("recent_chat_lines");
+	recent_lines = MYMIN(20, MYMAX(2, recent_lines));
```
Comment (paramat) [inline src/chat.cpp]: `std::min`, `std::max` or even better `rangelim(variable, lower limit, upper limit)`

File: builtin/settingtypes.txt (hunk 10 above)
Comment (SmallJoker) [reply builtin/settingtypes.txt]: If changing, then to "messages", as they may exceed one line depneding on font size, screensize and character limit.

Comment (paramat) [reply builtin/settingtypes.txt]: Ah, maybe 'messages' is better than 'lines' then?

**Comment by paramat** [comment]
Maybe 'messages' is better than 'lines' then for the setting name?

**Comment by nerzhul** [comment]
Merged as 4 squashed commits

5e24ca05fd91a7870983858f02b4f6cf7b496a06
792e013eaf4a572364457b96f6f9c07dafe6fe9a
93d185ee5d481c007d7b5569d046471dc60d40b6
0e8ee84d7477416dce4ad0d3fdde893676a9df6f
